Lib/asyncio/unix_events.py

```diff
@@ -308,6 +309,74 @@ def _child_watcher_callback(self, pid, returncode, transp):
                             ssl_handshake_timeout=ssl_handshake_timeout)
         return server
 
+    async def _sock_sf_fast(self, sock, file, offset=0, count=None):
+        try:
+            os.sendfile
+        except AttributeError as exc:
+            raise RuntimeError("os.sendfile() in not available")
+        fileno, blocksize = socket._prepare_sendfile(file, count)
+        if not blocksize:
+            return 0  # empty file
+
+        fut = self.create_future()
+        fd = sock.fileno()
+        self._sock_sf_fast_impl(fut, None, fd, fileno,
+                                offset, count, blocksize, 0)
+        return await fut
+
+    def _sock_sf_fast_impl(self, fut, registered_fd, fd, fileno, offset,
+                           count, blocksize, total_sent):
+        if registered_fd is not None:
+            # Remove the callback early.  It should be rare that the
+            # selector says the fd is ready but the call still returns
+            # EAGAIN, and I am willing to take a hit in that case in
+            # order to simplify the common case.
+            self.remove_writer(registered_fd)
+        if fut.cancelled():
```

Comment:
You copied this check from other `loop.sock_*` methods, but in fact it's incorrect.  It will only handle a case when a `Future` was cancelled right away.

The correct way of doing this (and that's btw what I'm doing in uvloop and planned to fix in asyncio) is to register a special callback on `fut`.  That callback should check if the future is cancelled, and if it is, cancel the sendfile operation (remove writer).

Otherwise, we can't cancel sendfile while `fd` is in selector.

Reply:
do we still need this check?